### DISCUSSION TITLE
Proposal to homogenize the examples

### DISCUSSION THREAD

**Discussion by ampuEus** [FreeOpcUa/opcua-asyncio#751] [General]
I have been using OPC UA for some time and I have seen that the examples do not have a homogeneous structure and a short introductory description explaining what is searched for each example (apart from the file name itself).

As a new user, it took me a bit to get used to them and my idea with this post is to see if you would like me to review the examples so that they all have a structure as similar as possible and with a small description so new users have it easier.

For example:
 - Sometimes, the "logging" is configured at the beginning of the program after importing the modules and other times in between or `_logger.info ('Starting server!')` Is sometimes set and sometimes not
 - Sometimes the formula `asyncio.run (main ())` is used and other times `asyncio.run (main (), debug = True)`
 - The certificates and the programs to generate them could be in a folder instead of mixed with other examples that do not use them
 - etc.

So If you want I can open a pull request and start making these little changes to see if they seem good to you.

P.S.: Since I have been using this library, I would like to contribute something. But I know that my level today is not up to being able to implement the OPC UA standard (alarms, XML protocol, session restore...). Even so, I think I could take away simpler jobs like this one that I'm proposing. Or some other task that, even though it is easy, you do not have the time to do it.

**Comment by oroulet** [comment]
you are welcome! starting with the example is a good idea. a merge request is more than welcome.
Maybe you should write down here the main things you want to change. 

also 

* logging MUST be configured AFTER "if __name__ == "__main__"

**Reply by ampuEus** [reply]
My idea is to start with these changes:

**1. Sometimes for the same piece of code each example has a different comment**
For example:
```
    # setup our own namespace, not really necessary but should as spec
    uri = 'http://examples.freeopcua.github.io'
    idx = await server.register_namespace(uri)
```
Sometimes it only appears `# setup our own namespace` and other times it doesn't even appear.

**2. Put all certificates in the same folder and the same with the xml**
My directory proposal:
```
 - examples
    - certificates
         - peer-certificate-example-1.der
         - peer-certificate-example-2.der
         - peer-certificate-example-3.der
         - peer-certificate-example-encrypted-private-key.der
         - peer-certificate-example-1.pem
         - peer-certificate-example-2.pem
         - peer-certificate-example-3.pem
         - peer-certificate-example-encrypted-private-key.der
         - certificate-3072-example.der
         - certificate-example.der
         - generate_certificate.sh
         - private-key-3072-example.pem
         - private-key-example.pem
  - xml custom models
         - custom_nodes.xml
         - customobject.xml
```
Of course, I have to see what parts of the code use these files and change their address to this new one.

**3. Put the logger in the right place**
**4. Use this structure for all the examples:**
  - Describe the purpose of the example
  - Import non-asyncua modules
  - Import asyncua modules
  - Example classes (If it has them)
  - Example functions (If it has them)
  - The main function
         - loggger
         - server setup
         - namespace
         - [specific code of the example]
         - start server
  - And the run
```
if __name__ == "__main__":
         logging.basicConfig(level=logging.INFO)
         asyncio.run(main())
```
